### ISSUE TITLE
Support Pagination

### ISSUE THREAD

**Issue by grsouza** [mergesort/Bodega#14]
Hi @mergesort I've implemented a CloudKit backend for Bodega here (https://github.com/binaryscraping/cloudkit-bodega).

But as it's remote storage, there are some non-ideal parts, such as the fetch of all data and keys, as depending on the amount of data, can be a huge problem.

For CloudKit, I've implemented this by looping through the cursors until all items are fetches, but ideally would be better if Bodega provided some way of paginating the results, then the caller could request the next page whenever it needs, and we don't need to fetch everything upfront.

```swift
 private func fetchAllRecords() async throws -> [(CKRecord.ID, Result<CKRecord, Error>)] {
    let query = CKQuery(recordType: recordType, predicate: NSPredicate(value: true))
    var results: [(CKRecord.ID, Result<CKRecord, Error>)] = []
    var (matchResults, queryCursor) = try await privateDatabase.records(matching: query)
    results.append(contentsOf: matchResults)

    while let cursor = queryCursor {
      (matchResults, queryCursor) = try await privateDatabase.records(continuingMatchFrom: cursor)
      results.append(contentsOf: matchResults)
    }

    return results
  }
```

**Comment by mergesort** [comment]
Hey @grsouza, this looks like a really neat StorageEngine use case to build atop Bodega, I'm excited to dive further into your code and see what you've made! I want to mention that between work, the holidays, and some family matters, it may take me some time to implement any requests.

This seems like a perfectly reasonable ask to me, though I will admit I'm not sure the best approach to take in a way that would scale to a general purpose API.

1. If you're not using Boutique you can always add more functions to your `CloudKitStorageEngine`, one that handles paginated requests. The downside of course is that Boutique doesn't have support for any functions that aren't in the protocol, but if you're not using Boutique then this should be ok.
2. I could add a `fetch(pagination: PaginationOptions)` method to the `StorageEngine` protocol, allowing the specification of pagination options. This could be tricky because many APIs support pagination, but they have different names, parameters, and ways to go about that.
3. I could create a `RemoteStorageEngine` protocol which adds the pagination options, which would require a well-defined API in mind before I do that.

I'd love to know if you've had any thoughts of how you'd like to see this problem solved, what a good API would look like to you. Right now my mind is a bit of a blank slate but judging by your repo you've been thinking about this idea for at least a week or two.

---

I also would like to know [the project](https://github.com/binaryscraping/cloudkit-bodega) you've pushed, is it meant to be for personal use or a more generalizable solution? I have some decent experience with CloudKit and noticed a lot of assumptions (for example using the `private` database, which is perfectly fine assuming it's meant to be a solution tailored to your needs rather than a more general `CloudKitStorageEngine`.

**Comment by grsouza** [comment]
Hey @mergesort thanks for the quick response.

I'm not sure yet what my goal is with this library if it is for personal use only or a more general purpose, I don't have much experience with CloudKit, and I'd need some help to make it more general.

---

For the pagination, the pagination can also be used on other engines that are not remote, for example, the SQLite one, even tho SQLite is pretty fast for fetching all data, having a pagination option would be great.

I thought of using an `AsyncSequence` as the return type of the paginated method, that way the caller can keep calling its `next()` method until it fetches all pages and each `next` call returns a page of results.

**Comment by mergesort** [comment]
This is a really cool implementation @grsouza! I hadn't considered using `AsyncSequence`, but now that I've read through the proposal it seems like the perfect choice in retrospect.

I was having issues pushing updates to your pull request and accidentally created a new branch, branching off of your proposal. https://github.com/mergesort/Bodega/tree/binaryscraping/main

I don't really have many usability changes in mind, but I pushed up a few changes to the proposal that I'd appreciate your feedback on.

- I renamed `RemoteStorageEngine` to `PaginatedStorageEngine` since the protocol carves out a distinction between a paginated and unpaginated `StorageEngine`, rather whether the `StorageEngine` is local or remote as we'd initially discussed with your `CloudKitStorageEngine` idea.
- I made the `PaginationOptions` limit non-optional, since I think if you're opting into creating a `PaginatedStorageEngine` you should have to specify a limit. I may have misunderstood why this property was optional so if I did please let me know and we can change it back.

There are to things I'd love to do before merging this in.
1. I'd love to try this out in a real project, to get a feel for how it's used and to see if there are any limitations of the API, or if improvements are necessitated in realistic usage.
2. `PaginatedStorageEngine` would need to be documented, both inline and as part of a docc article for people coming to the project without as much context as you and I have.

I also have two open questions.
1. Do you have any thoughts on the best way to integrate this into [Boutique](https://github.com/mergesort/Boutique)? It seems like a no-brainer that a developer could want to display the contents of a `Store` in a paginated fashion, so limiting this to Bodega isn't living up to the full potential of what we've got there. At the moment all of the assumptions in Boutique are that you're loading the full contents of your `Store`, but I'd be open to revisiting that with this new pagination functionality in place.
2. Should we provide other `PaginationOption` types? For example an application like a Twitter client would have a Twitter Status ID based cursor to keep timelines consistent if a user deletes a tweet. We don't have that here, and it could definitely be a premature optimization, but I'd like to consider that option when thinking about the way pagination is represented to a `PaginatedStorageEngine`.

Thank you again for the proposal, it's a great idea, and I especially appreciate the implementation. The implementation helped make this idea a lot more tangible, and now I can definitely see a path for this functionality going into Bodega.

**Comment by mergesort** [comment]
This has now been merged in #23, without any special integration into Boutique. I didn't add any `PaginationOption` types, since it's an additive change I figure it can always be added later if needed.

Thank you again @grsouza for the idea, the suggestions, changes, and for making this happen!
